google-cloud-firestore/lib/google/cloud/firestore/database.rb

```diff
@@ -0,0 +1,1013 @@
+        end
+
+        ##
+        # Create a transaction to perform multiple reads and changes that are
+        # executed atomically at a single logical point in time in a database.
+        #
+        # All changes are accumulated in memory until the block completes.
+        # Transactions will be automatically retried when possible. See
+        # {Transaction}.
+        #
+        # @see https://firebase.google.com/docs/firestore/manage-data/transactions
+        #   Transactions and Batched Writes
+        #
+        # @param [String] previous_transaction The transaction identifier of a
+        #   transaction that is being retried. Read-write transactions may fail
+        #   due to contention. A read-write transaction can be retried by
+        #   specifying `previous_transaction` when creating the new transaction.
+        #
+        #   Specifying `previous_transaction` provides information that can be
+        #   used to improve throughput. In particular, if transactional
+        #   operations A and B conflict, specifying the `previous_transaction`
+        #   can help to prevent livelock. (See {Transaction#transaction_id})
+        #
+        # @yield [transaction] The block for reading data and making changes.
+        # @yieldparam [Transaction] transaction The transaction object for
+        #   making changes.
+        #
+        # @return [Time] The time the changes were committed
+        #
+        # @example
+        #   require "google/cloud/firestore"
+        #
+        #   firestore = Google::Cloud::Firestore.new
+        #
+        #   firestore.transaction do |tx|
+        #     # Set the data for NYC
+        #     tx.col("cities").doc("NYC").set({ name: "New York City" })
+        #
+        #     # Update the population for SF
+        #     tx.col("cities").doc("SF").update({ population: 1000000 })
+        #
+        #     # Delete LA
+        #     tx.col("cities").doc("LA").delete
+        #   end
+        #
+        def transaction previous_transaction: nil
+          retries = 0
+          backoff = 1.0
+          deadline = 60
+          start_time = Time.now
+
+          transaction = Transaction.from_database \
+            self, previous_transaction: previous_transaction
+          begin
+            yield transaction
+            transaction.commit
+          rescue Google::Cloud::UnavailableError => err
+            # Re-raise if deadline has passed
+            raise err if Time.now - start_time > deadline
+            # Sleep with incremental backoff
+            sleep(backoff *= 1.3)
+            # Create new transaction and retry
+            transaction = Transaction.from_database \
+              self, previous_transaction: transaction.transaction_id
+            retries += 1
+            retry
+          rescue Google::Cloud::InvalidArgumentError => err
+            # Return if a previous call has succeeded
+            return nil if retries > 0
```

Comment:
Does this block automatically cause retries? We need to make sure we retry on commit failures without backoff when a transaction fails due to a document changes on the backend.